### ISSUE TITLE
Support styling libraries via adapters

### ISSUE THREAD

**Issue by asyarb** [asyarb/babel-plugin-style-props#10]
Right now, this babel plugin has pretty tight integration with `styled-components` and `emotion`. Logic for supporting these libraries is sprinkled around our various builders, and it would be nice to at least isolate them to dedicated functions/adapters. Generally, these adapters/functions would run  _after_ we build our theme-aware `css` prop.

Approaching the plugin in this manner will make it easier for us to extend it's capabilities as future CSS-in-JS solutions become popular, and/or make it easier for contributors to add support for their respective solution.

---

Ideally our general flow would be something like this:

1. Gather and strip style props from visited JSX element.
2. For each style prop, build our theme aware expression as an object property.
3. Build our new `css` prop with merging if necessary
4. Run our new `css` prop through an adapter, which transforms it to adhere to the styling library it's associated with.
5. Apply the prop to the JSX element.

Right now, `4` happens in scattered locations when `2` is happening.

**Comment by asyarb** [comment]
This roadmap has been revised, and is summarized below:

---

To make this babel plugin more reusable, we've decided to take a holistically different approach that is more framework agnostic.

Going forward, this plugin will **only** parse, themify, and optionally strip style props and place them in a simple name-spaced object. It will do this agnostically of any styling library restrictions and adhere to a standard convention. 

**Doing this causes three things:**

1. It allows developers to create reusable libraries that rely on this plugin. With our current approach, any library trying to abstract reusable generics can't run their code through babel since it would hard-code their breakpoints to whatever the library author specifies.

This is unideal, since most of the time we want that kind of control to be in the consumer of such a library.

2. It allows us and any future contributors to create adapters that work with this known name-spaced object. 

This effectively isolates  library specific logic in dedicated plugins that run **after** this plugin by looking for this name-spaced object and doing anything it wants with it. This opens the door for all CSS-in-JS solutions, full static extraction, partial extraction, and basically anything in-between.

3. It makes this plugin itself easier to maintain. Hooray!

**Comment by asyarb** [comment]
Progress on this rewrite can be tracked in the `rewrite` branch.

**Comment by asyarb** [comment]
For personal reference:

Styleprops will be shaped as follows:

```jsx
<div
      __styleProps__={[
        {
          css: {
            base: [
              {
                borderBottom: '1px solid',
              },
              {},
              {
                margin: '3rem',
              },
            ],
            hover: [],
            focus: [],
            active: [],
          },

          extensions: {
            scales: {
              margin: size,
              padding: 'l',
              border: ['xl', null, 'l'],
            },
          },
        },
        ,
      ]}
    />
```

**Comment by asyarb** [comment]
Released 0.3.1 which features this new model.
